rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,813 @@
++ return {
++   root: mergeClasses(
++     accordionHeaderClassNames.root,
++     styles.root,
++     state.inline && styles.rootInline,
++     state.disabled && styles.rootDisabled,
++     state.root.props.className,
++   );
++   icon: state.icon && mergeClasses(
++     accordionHeaderClassNames.icon,
++     styles.icon,
++     state.icon.props.className
++   )
++ }
+};
+
+// AccordionHeader
+export const AccordionHeader: ForwardRefComponent<AccordionHeaderProps> = React.forwardRef((props, ref) => {
+  const state = useAccordionHeader_unstable(props, ref);
+- useAccordionHeaderStyles_unstable(state);
++ const styles = useAccordionHeaderStyles_unstable(state);
+  const contextValues = useAccordionHeaderContextValues_unstable(state);
+- return renderAccordionHeader_unstable(state, contextValues);
++ return renderAccordionHeader_unstable(state, styles, contextValues);
+});
+
+// renderAccordionHeader
+- export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
++ export const renderAccordionHeader_unstable = (state: AccordionHeaderState, styles: AccordionHeaderStyles) => (
+-  <state.root>
++  <state.root className={styles.root}>
+    <state.button>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+##### Passing overrides from state to render when needed
+
+Since our logic lies on the state hooks, the override methods will be implemented on those state hooks, although it should be
+declared on the render method explicitly to properly work as an override.
+
+For example, on the case `AccordionHeader` the `button` slot would override the `onClick` method. For that a property called `overrides`
+can be explicitly added to `AccordionHeaderState` stating such override:
+
+```tsx
+// AccordionHeader.types.ts
+type AccordionHeaderState = NextComponentState<AccordionHeaderSlots> & {
+  overrides: {
+    button: {
+      onClick(ev: React.MouseEvent<HTMLButtonElement>): void;
+    };
+  };
+};
+
+// useAccordionHeader.ts
+const state = {
+  // ...
+  overrides: {
+    button: {
+      onClick: useEventCallback(ev => {
+        buttonSlot.props.onClick?.(ev);
+        if (!ev.defaultPrevented) {
+          onAccordionHeaderClick(ev);
+        }
+      }),
+    },
+  },
+};
+
+// renderAccordionHeader.tsx
+const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
+  <state.root>
+    <state.button onClick={state.overrides.button.onClick}>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+#### Custom Pragma implementation
```

Comment:
> I think we'd need to have a way to override props for every slot, built-in:
We'll always need to override className.
Someone re-composing our component may want to override props for a slot that we don't override internally.

Totally doable, we could include a `state.overrides.SLOT` as part of the proposal

Reply:
> I think this idea has good merit, and I think we can do without a custom pragma. The JSX pragma is just syntactic sugar, but if we simplify the non-JSX syntax enough, we could skip using JSX in the render functions entirely.
> 
> Here's an idea that might work ok without needing a JSX pragma. A few background assumptions:
> 
> * The only job of the `renderFoo` function is to define the _tree structure_ of the control (i.e. what slots go where in the DOM tree); not to override props or add functionality.
> * We never should have property overrides in `renderFoo`: they should all be handled by the `useFoo` hook(s).
> * All of our render functions adhere to this, and don't provide custom prop overrides in the render function.
> 
> With those assumptions, we could then write a `renderSlot` function that takes the slot object, and children:
> 
> ```ts
> function renderSlot<Props extends UnknownSlotProps>(
>   component: SlotComponent<Props>, 
>   children: React.ReactNode[]
> ): React.ReactElement<Props> | null {
>    // Works just like jsxFromSlotComponent, except without override handling
> }
> ```
> 
> Then a render function might look like:
> 
> ```ts
> const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
>   renderSlot(state.root, [
>     renderSlot(state.button, [
>       state.expandIconPosition === 'start' && state.expandIcon && renderSlot(state.expandIcon),
>       state.icon && renderSlot(state.icon),
>       ...state.root.props.children,
>       state.expandIconPosition === 'end' && state.expandIcon && renderSlot(state.expandIcon),
>     ]),
>   ]);
> );
> ```
> 
> ... a little old school like calling React.createElement manually. The main distinction is that by preventing further overrides of props, the function doesn't become a mess of props mixed with children: it _only_ defines the structure of the slots via the children array.

This is just going back to the way things are done in N*. I believe a custom pragma is a better way forward for readability

Reply:
> We could still wrap all this with a custom JSX pragma if we wanted to. I'm not strongly opposed to it, although it has some downsides: namely developer confusion about why our JSX works differently from others. We could consider whether or not to add a JSX pragma as a separate topic.
> 
> But for reference, with a custom pragma:
> 
> ```ts
> function jsxRenderSlot(
>   component: SlotComponent<Props>,
>   overrides?: never, // prevent any overrides from being added -- not sure if this is the right type
>   ...children: React.ReactNode[]
> ): {
>   return renderSlot(component, children);
> }
> ```
> 
> The result would look more similar to your other examples, except you couldn't specify any extra props on the nodes.
> 
> ```ts
> const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
>   <state.root>
>     <state.button>
>       {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
>       {state.icon && <state.icon />}
>       {state.root.props.children}
>       {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
>     </state.button>
>   </state.root>
> );
> ```

Yeah, that's not the right types but I got what is your idea. So in this ase we would completely avoid overrides from the render methods and we would need to introduce it solely on the state hooks, right?! @behowell, I'm not opposed to this idea, it could blend in just fine with option A, C, D

Reply:
> The custom JSX pragma could be confusing to other developers, since it is a rare and unexpected thing to do.

There is adoption in the community, as it's one of options to use `css` prop with Emotion: https://emotion.sh/docs/css-prop#jsx-pragma

Reply:
> Just wanted to summarize my opinion here:
> 
> * A custom JSX pragma is not _required_ for the changes in this RFC. The changes in this RFC can be implemented with a function like `renderSlot`.

Wasn't this clear already? option B is basically the same idea as this `renderSlot` but focused in the children problem, the main difference is that this `renderSlot` approach just goes way back to the way things were done in N*.

> * IMO the function-only approach is _only slightly_ less readable than the JSX pragma.

I disagree completely, and I believe the whole idea behind the creation of JSX syntax also disagrees with this argument 😅. 

> * We could optionally add the JSX pragma later if we decide to.

Yes, that would be something similar to option B. I guess my point with this RFC was proposing changes **together**
with the solution of the problem. I guess I didn't emphasise that well, I'll try to do it better.